### PR TITLE
fix(Scripts/Midsummer): Implement Ribbon Pole dance effects

### LINKED ISSUE #12145: Midsummer pole dance effects missing
Original Chromiecraft report https://github.com/chromiecraft/chromiecraft/issues/3647 

### What client do you play on?

enUS

### Faction

- [X] Alliance
- [X] Horde

### Content Phase:

- [x] Generic
- [ ] 1-19
- [ ] 20-29
- [ ] 30-39
- [ ] 40-49
- [ ] 50-59
- [ ] 60

### Current Behaviour

Ribbon attaches to the bottom of the pole
All the effects depending on the amount of dancing people are missing

### Expected Blizzlike Behaviour

Ribbon attaches at the top
1 person dancing adds a blue ring of fire
2 add fireworks
3 add ground flowers
6 makes it spew lava
7 people make big dancing flames spawn (npc entry 26267)

### Source

https://www.wowhead.com/object=181605/ribbon-pole#comments:id=312207
https://www.wowhead.com/npc=26267/big-dancing-flames
https://youtu.be/f4PwcnOjTCg

### Steps to reproduce the problem

1. Have 7 people use the pole
2. Notice no effects, and ribbon is on the base of the pole

### AC rev. hash/commit

https://github.com/chromiecraft/azerothcore-wotlk/commit/82cfd1b8d4147f09564e692f13effc4142178d72

### Operating system

Ubuntu 20.04

### Modules

- [mod-ah-bot](https://github.com/azerothcore/mod-ah-bot)
- [mod-bg-item-reward](https://github.com/azerothcore/mod-bg-item-reward)
- [mod-cfbg](https://github.com/azerothcore/mod-cfbg)
- [mod-chat-transmitter](https://github.com/azerothcore/mod-chat-transmitter)
- [mod-chromie-xp](https://github.com/azerothcore/mod-chromie-xp)
- [mod-cta-switch](https://github.com/azerothcore/mod-cta-switch)
- [mod-desertion-warnings](https://github.com/azerothcore/mod-desertion-warnings)
- [mod-dmf-switch](https://github.com/azerothcore/mod-dmf-switch)
- [mod-duel-reset](https://github.com/azerothcore/mod-duel-reset)
- [mod-eluna](https://github.com/azerothcore/mod-eluna)
- [mod-ip-tracker](https://github.com/azerothcore/mod-ip-tracker)
- [mod-low-level-arena](https://github.com/azerothcore/mod-low-level-arena)
- [mod-low-level-rbg](https://github.com/azerothcore/mod-low-level-rbg)
- [mod-multi-client-check](https://github.com/azerothcore/mod-multi-client-check)
- [mod-progression-system](https://github.com/azerothcore/mod-progression-system)
- [mod-pvp-titles](https://github.com/azerothcore/mod-pvp-titles)
- [mod-pvpstats-announcer](https://github.com/azerothcore/mod-pvpstats-announcer)
- [mod-queue-list-cache](https://github.com/azerothcore/mod-queue-list-cache)
- [mod-rdf-expansion](https://github.com/azerothcore/mod-rdf-expansion)
- [mod-transmog](https://github.com/azerothcore/mod-transmog)
- [mod-weekend-xp](https://github.com/azerothcore/mod-weekend-xp)
- [lua-carbon-copy](https://github.com/55Honey/Acore_CarbonCopy)
- [lua-exchange-npc](https://github.com/55Honey/Acore_ExchangeNpc)
- [lua-custom-worldboss](https://github.com/55Honey/Acore_CustomWorldboss)
- [lua-level-up-reward](https://github.com/55Honey/Acore_LevelUpReward)
- [lua-recruit-a-friend](https://github.com/55Honey/Acore_RecruitAFriend)
- [lua-send-and-bind](https://github.com/55Honey/Acore_SendAndBind)
- [lua-temp-announcements](https://github.com/55Honey/Acore_TempAnnouncements)
- [lua-zonecheck](https://github.com/55Honey/acore_Zonecheck)
- [lua-zone-debuff](https://github.com/55Honey/Acore_ZoneDebuff)

### Customizations

None

### Server

ChromieCraft

### PR DESCRIPTION
**Pull request by sudlud**
## Changes Proposed:
This PR proposes changes to:
-  [ ] Core (units, players, creatures, game systems).
-  [x] Scripts (bosses, spell scripts, creature scripts).
-  [x] Database (SAI, creatures, etc).

If your pull request promotes complex changes that require a detailed explanation, please describe them in detail specifying what your solution is and what is it meant to address.

## Issues Addressed:

- Closes #12145
- Closes chromiecraft/chromiecraft#3647

## SOURCE:

The changes have been validated through:
- [ ] Live research (checked on live servers, e.g Classic WotLK, Retail, etc.)
- [ ] Sniffs (remember to share them with the open source community!)
- [x] Video evidence, knowledge databases or other public sources (e.g forums, Wowhead, etc.)
- [ ] The changes promoted by this pull request come partially or entirely from another project (cherry-pick). **Cherry-picks must be committed using the proper --author tag in order to be accepted, thus crediting the original authors, unless otherwise unable to be found**

## Tests Performed:

This PR has been:
- [x] Tested in-game by the author.
- [ ] Tested in-game by other community members/someone else other than the author/has been live on production servers.
- [ ] This pull request requires further testing and may have edge cases to be tested.

## How to Test the Changes:

- [ ] This pull request can be tested by following the reproduction steps provided in the linked issue
- [ ] This pull request requires further testing. Provide steps to test your changes. If it requires any specific setup e.g multiple players please specify it as well.

.event start 1
.tele orgrimmar
.go xyz 1937 -4326 22
click on ribbon pole

observe effects depending on dancing player count:

1 random blue / red ring of fire
-> 50% chance, updates every 5s

2 add fireworks (1 rocket)
-> 15s interval

3 add ground flowers
-> 10s interval, max. 3

3 upgrade fireworks to 3 rockets

5 upgrade fireworks to 5 rockets

6 makes it spew lava
-> choosing 2 random targets every 10 seconds

7 people make big dancing flames spawn
-> 30s interval, max. 4, despawn after 5min

If a dancing player leaves the pole, the effects get reduced to the previous level.

The thresholds, timings and max. counts of spawned NPCs can be modified in the enum RibbonPole, e.g. setting all THRESHOLD_...'s to 1 will trigger all effects for 1 dancing player for testing purposes.

## Known Issues and TODO List:

- [ ]
- [ ]

## How to Test AzerothCore PRs
 
When a PR is ready to be tested, it will be marked as **[WAITING TO BE TESTED]**.

You can help by testing PRs and writing your feedback here on the PR's page on GitHub. Follow the instructions here:

http://www.azerothcore.org/wiki/How-to-test-a-PR

**REMEMBER**: when testing a PR that changes something **generic** (i.e. a part of code that handles more than one specific thing), the tester should not only check that the PR does its job (e.g. fixing spell XXX) but **especially** check that the PR does not cause any regression (i.e. introducing new bugs).

**For example**: if a PR fixes spell X by changing a part of code that handles spells X, Y, and Z, we should not only test X, but **we should test Y and Z as well**.

### CODE REVIEW

**Comment by sudlud** [comment]
Spell script is missing validate(), I will add it later today

File: src/server/scripts/Events/midsummer.cpp [hunk 1]
```diff
@@ -210,15 +211,331 @@ class spell_gen_crab_disguise : public AuraScript
...
+        if (!_bunny)
+            return;
+
+        if (_dancerList.size() >= THRESHOLD_FIREWORK)
+        {
+            _bunny->CastSpell(nullptr, SPELL_RIBBON_POLE_FIREWORK);
+        }
+        if (_dancerList.size() >= THRESHOLD_FIREWORK_3)
+        {
+            _scheduler.Schedule(Milliseconds(500), [this](TaskContext /*context*/)
```
Comment (Kitzunu) [inline src/server/scripts/Events/midsummer.cpp:427]: can be 500ms

File: src/server/scripts/Events/midsummer.cpp [hunk 2]
```diff
@@ -210,15 +211,331 @@ class spell_gen_crab_disguise : public AuraScript
...
+    Creature* _bunny;
+};
+
+class spell_midsummer_ribbon_pole_firework : public SpellScript
+{
+    PrepareSpellScript(spell_midsummer_ribbon_pole_firework)
+
+    void ModDestHeight(SpellDestination& dest)
+    {
+        Position const offset = { 0.0f, 0.0f, 20.0f , 0.0f};
```
Comment (Kitzunu) [inline src/server/scripts/Events/midsummer.cpp:531]: spacing after 0.0f

**Comment by sudlud** [comment]
@Kitzunu i thinks his should be it for now.
With this PR the midsummer ribbonPole finally shows some action!

Im sure that this could still be improved in the future with sniffs / observations to get the timings etc. as blizzlike as possible.

File: src/server/scripts/Events/midsummer.cpp [hunk 3]
```diff
@@ -210,15 +211,290 @@ class spell_gen_crab_disguise : public AuraScript
...
+        if (!_bunny)
+            return;
+
+        if (_dancerList.size() >= THRESHOLD_SPEW_LAVA)
+        {
+            if (!_dancerList.empty())
+            {
+                std::vector<Player*>::iterator begin = _dancerList.begin(), end = _dancerList.end();
+                std::random_device rd;
+                std::shuffle(begin, end, std::default_random_engine{ rd() });
```
Comment (Winfidonarleyan) [inline src/server/scripts/Events/midsummer.cpp:439]: https://github.com/azerothcore/azerothcore-wotlk/blob/366b8d01fc0c90d97a922893aee03c5010418867/src/common/Utilities/Containers.h#L135

Comment (sudlud) [reply src/server/scripts/Events/midsummer.cpp:439]: Thanks for the hint, will update the code to use utilities

File: data/sql/updates/pending_db_world/rev_1696077614656614200.sql [hunk 4]
```diff
@@ -0,0 +1,3 @@
+-- midsummer creatureAI npc_midsummer_ribbon_pole_target
+UPDATE `creature_template` SET `AIName` = 'SmartAI' WHERE `entry` = 17066;
+UPDATE `creature_template` SET `ScriptName` = 'npc_midsummer_ribbon_pole_target' WHERE `entry` = 17066;
```
Comment (Kitzunu) [inline data/sql/updates/pending_db_world/rev_1696077614656614200.sql:3]: You cant have both an AIName and ScriptName set.

Comment (Kitzunu) [reply data/sql/updates/pending_db_world/rev_1696077614656614200.sql:3]: Also you could combine all SQL files into one file.

File: src/server/scripts/Events/midsummer.cpp [hunk 5]
```diff
@@ -210,15 +211,288 @@ class spell_gen_crab_disguise : public AuraScript
...
+    THRESHOLD_SPEW_LAVA                 = 6,
+    THRESHOLD_DANCING_FLAMES            = 7,
+
+    // update interval in Seconds(...)
+    INTERVAL_CLEANUPCHECKS              = 1,
+    INTERVAL_FLAME_CIRCLE               = 5,
+    INTERVAL_FIREWORK                   = 15,
+    INTERVAL_GROUND_FLOWERS             = 10,
+    INTERVAL_SPEW_LAVA                  = 10,
+    INTERVAL_DANCING_FLAMES             = 15,
```
Comment (Kitzunu) [inline src/server/scripts/Events/midsummer.cpp:254]: this is unnecessary

File: src/server/scripts/Events/midsummer.cpp [hunk 6]
```diff
@@ -210,15 +211,288 @@ class spell_gen_crab_disguise : public AuraScript
...
+
+        if (_dancerList.size() >= THRESHOLD_FIREWORK)
+        {
+            _bunny->CastSpell(nullptr, SPELL_RIBBON_POLE_FIREWORK);
+        }
+        if (_dancerList.size() >= THRESHOLD_FIREWORK_3)
+        {
+            _scheduler.Schedule(500ms, [this](TaskContext /*context*/)
+            {
+                if (_bunny)
```
Comment (Kitzunu) [inline src/server/scripts/Events/midsummer.cpp:389]: no need. You already return if _bunny dont exist higher up

File: data/sql/updates/pending_db_world/rev_1696336524256958400.sql [hunk 7]
```diff
@@ -0,0 +1,6 @@
+-- midsummer NPC Big Dancing Flame - SmartAI - fire dance spell
+UPDATE `creature_template` SET `AIName` = 'SmartAI' WHERE `entry` = 26267;
+
+DELETE FROM `smart_scripts` WHERE (`source_type` = 0 AND `entryorguid` = 26267);
+INSERT INTO `smart_scripts` (`entryorguid`, `source_type`, `id`, `link`, `event_type`, `event_phase_mask`, `event_chance`, `event_flags`, `event_param1`, `event_param2`, `event_param3`, `event_param4`, `event_param5`, `event_param6`, `action_type`, `action_param1`, `action_param2`, `action_param3`, `action_param4`, `action_param5`, `action_param6`, `target_type`, `target_param1`, `target_param2`, `target_param3`, `target_param4`, `target_x`, `target_y`, `target_z`, `target_o`, `comment`) VALUES
+(26267, 0, 0, 0, 54, 0, 100, 0, 0, 0, 0, 0, 0, 0, 11, 45418, 0, 0, 0, 0, 0, 1, 0, 0, 0, 0, 0, 0, 0, 0, 'spell cast - NPC Big Dancing Flame Fire Dance');
```
Comment (Kitzunu) [inline data/sql/updates/pending_db_world/rev_1696336524256958400.sql:6]: fix the comment. Should be `owner - event - action`

Comment (Kitzunu) [reply data/sql/updates/pending_db_world/rev_1696336524256958400.sql:6]: for example `Goblin - OOC - Cast Spell 'Spell Name'`

File: src/server/scripts/Events/midsummer.cpp [hunk 8]
```diff
@@ -210,15 +211,288 @@ class spell_gen_crab_disguise : public AuraScript
...
+    SPELL_RIBBON_POLE_PERIODIC_VISUAL   = 45406,
+    // spew lava trails
+    SPELL_RIBBON_POLE_SPEW_LAVA_VISUAL  = 45421,
+    // blue fire ring, duration 5s
+    SPELL_FLAME_RING                    = 46842,
+    // red fire ring, duration 5s
+    SPELL_FLAME_PATCH                   = 46836,
+    // single firework explosion
+    SPELL_RIBBON_POLE_FIREWORK          = 46847,
+    SPELL_RIBBON_POLE_GROUND_FLOWER     = 46969,
```
Comment (Kitzunu) [inline src/server/scripts/Events/midsummer.cpp:231]: Make sure they are the actual spellname

**Review by Kitzunu — APPROVED** [review]

**Comment by TimekillerTK** [comment]
Tested. Ran 8 clients to have 8 characters to test the behaviour.

---

### Before (on master):
- Ribbon attaches to **the top** of the pole
- Each player spawns a ribbon and spins around in flames
- Regardless of the amount of ribbons, behavior stays the same

image


---

### After (on PR):
> 1 random blue / red ring of fire
> -> 50% chance, updates every 5s

Observed as specified (only red flames pictured, but it was rotating between red and blue):

image


> 2 add fireworks (1 rocket)
> -> 15s interval

Observed as specified.

image

> 3 add ground flowers
> -> 10s interval, max. 3
> 
> 3 upgrade fireworks to 3 rockets
> 

Observed as specified.

image

image

> 5 upgrade fireworks to 5 rockets

Observed as specified.

image


> 6 makes it spew lava
> -> choosing 2 random targets every 10 seconds

Observed as specified.

image

> 7 people make big dancing flames spawn
> -> 30s interval, max. 4, despawn after 5min

Observed as specified, **however, the dancing flames despawn every 1min** instead of 5min. Is that intended @sudlud ?

image


> If a dancing player leaves the pole, the effects get reduced to the previous level.

This was also observed as specified. I checked the behavior subtracting each player from the pole:
- 7 -> 6 Effect reduced, the dancing flames stick around until their timer ends, and they despawn naturally (After 1 min)
- 6 -> 5 Lava disappears
- 5 -> 4 Only 3 rockets & Ground flowers
- 4 -> 3 Same thing
- 3 -> 2 Ground Flowers Disappear, Only 1 rocket
- 2 -> 1 Only Random Blue Flame

**Comment by sudlud** [comment]
@TimekillerTK thanks for testing this!

**Comment by sudlud** [comment]
Is there anything to do from my side here atm? :)

**Pull request merged**